src/util/secrets/secrets.c

```diff
@@ -0,0 +1,1235 @@
+        ret = EIO;
+        goto done;
+    }
+
+    ret = lcl_read_mkey(sec_ctx, &sec_ctx->master_key);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Cannot get the master key\n");
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    if (ret == EOK) {
+        *_sec_ctx = talloc_steal(mem_ctx, sec_ctx);
+    }
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static int local_db_dn(TALLOC_CTX *mem_ctx,
+                       struct ldb_context *ldb,
+                       const char *basedn,
+                       const char *req_path,
+                       struct ldb_dn **req_dn)
+{
+    struct ldb_dn *dn;
+    const char *s, *e;
+    int ret;
+
+    dn = ldb_dn_new(mem_ctx, ldb, basedn);
+    if (!dn) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    s = req_path;
+
+    while (s && *s) {
+        e = strchr(s, '/');
+        if (e) {
+            if (e == s) {
+                s++;
+                continue;
+            }
+            if (!ldb_dn_add_child_fmt(dn, "cn=%.*s", (int)(e - s), s)) {
+                ret = ENOMEM;
+                goto done;
+            }
+            s = e + 1;
+        } else {
+            if (!ldb_dn_add_child_fmt(dn, "cn=%s", s)) {
+                ret = ENOMEM;
+                goto done;
+            }
+            s = NULL;
+        }
+    }
+
+    DEBUG(SSSDBG_TRACE_INTERNAL,
+          "Local path for [%s] is [%s]\n",
+          req_path, ldb_dn_get_linearized(dn));
+    *req_dn = dn;
+    ret = EOK;
+
+done:
+    return ret;
+}
+
+errno_t sss_sec_new_req(TALLOC_CTX *mem_ctx,
+                        struct sss_sec_ctx *sec_ctx,
+                        const char *url,
+                        uid_t client,
+                        struct sss_sec_req **_req)
+{
+    struct sss_sec_req *req;
+    TALLOC_CTX *tmp_ctx;
+    errno_t ret;
+
+    if (sec_ctx == NULL || url == NULL || _req == NULL) {
+        return EINVAL;
+    }
+
+    tmp_ctx = talloc_new(mem_ctx);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    req = talloc_zero(mem_ctx, struct sss_sec_req);
+    if (req == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+    req->sctx = sec_ctx;
+
+    ret = sss_sec_map_path(req, url, client, &req->mapped_path);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    /* drop the prefix and select a basedn instead */
+    if (strncmp(req->mapped_path,
+                SSS_SEC_BASEPATH, sizeof(SSS_SEC_BASEPATH) - 1) == 0) {
+
+        if (geteuid() != 0 && client != geteuid()) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "Only root can impersonate other users\n");
+            return EPERM;
+        }
+
+        req->path = talloc_strdup(req,
+                                     req->mapped_path + (sizeof(SSS_SEC_BASEPATH) - 1));
+        req->basedn = SECRETS_BASEDN;
+        req->quota = sec_ctx->quota_secrets;
+    } else if (strncmp(req->mapped_path,
+               SSS_SEC_KCM_BASEPATH, sizeof(SSS_SEC_KCM_BASEPATH) - 1) == 0) {
```

Comment:
I'd also move SSS_SEC_KCM_BASEPATH to align with req->mapped_path and the sizeof(...) to align with the SSS_SEC_KCM_BASEPATH

Reply:
fixed